### DISCUSSION TITLE
Case-sensitive replace in multi-selection

### DISCUSSION THREAD

**Discussion by bcspragu** [helix-editor/helix#6409] [Ideas]
Frequently, I'll have some block of code where I want to copy it and replace a certain word everywhere. For example, I might start with:

```go
if o.startTime != nil {
  opts = append(opts, somepkg.WithStartTime(o.startTime))
}
```

And I'll want to add a corresponding block for an `endTime` parameter. So I'll copy the block and do something like <kbd>Alt</kbd>+<kbd>o</kbd> <kbd>Alt</kbd>+<kbd>o</kbd> (to select the whole statement block) <kbd>s</kbd><kbd>start</kbd><kbd>c</kbd><kbd>end</kbd>. And I'll end up with:

```go
if o.endTime != nil {
  opts = append(opts, somepkg.WithendTime(o.endTime))
}
```

The problem is the `WithendTime`, which should naturally be `WithEndTime`. And for small cases, I just manually change it, but for larger/more complicated cases, it ends up being quite tedious, so I end up replacing each variant (`Start`, `start`, `START`, etc) individually.

I think the way I'd implement this idea is to have a 'case-sensitive' variant of <kbd>c</kbd> (maybe <kbd>Alt</kbd>+<kbd>c</kbd>?) that automatically substituted matches based on their case.

I saw an open issue (https://github.com/helix-editor/helix/issues/4978) which looked vaguely related, but I couldn't tell if that's advocating for the same thing.

**Comment by pascalkuthe** [comment]
I don't think this makes sense as a separate feature because you could only do this for the first character. For more complex casing rules this is impossible to future out for the editor.

I would suggest just doing multiple seperate replaces instead. It's quite easy to do. First replace all non-lowercase varatants (doesn't match the lowercase version because smart case) and then do the lowercase variant last (at that point canse insensitive won't matter). You could also force case-senstive mode by prefixing your search with ``(?-i)` but getting to order right requires less typing.  After each replace you can go back to your parent selection with `C-o`.

**Reply by bcspragu** [reply]
> I don't think this makes sense as a separate feature because you could only do this for the first character. For more complex casing rules this is impossible to future out for the editor.

For literal replacements (in my example "`start`"), one should be able to do it for all characters, though literals are definitely a special case, and even then it requires assorted special sub-cases. E.g. for `min(length(original), length(replacement))`, use the same case as the original, then extend the same case as the original after that. Definitely messy, but I imagine the majority of cases boil down to "replace one literal with another, being conscious of casing". Back in my Vim/NeoVim using days, [vim-abolish](https://github.com/tpope/vim-abolish#substitution), handled this type of stuff for me. I don't know what gnarly edge cases it handled behind the scenes, but I generally didn't have to think about it

> I would suggest just doing multiple seperate replaces instead. It's quite easy to do. [...]

This is exactly what I do now (noted in original post), and it's fine, but it's one of those (minor, nitty) things that takes you out of the flow of whatever you're doing. Once Helix supports plugins I might take my hand at making something here.
